api/services/retention/rollback_archived_workflow_run.py

```diff
@@ -0,0 +1,346 @@
+
+                # Verify row counts match manifest
+                manifest_total = sum(info.get("row_count", 0) for info in tables.values())
+                restored_total = sum(result.restored_counts.values())
+
+                if not self.dry_run:
+                    # Note: restored count might be less than manifest count if records already exist
+                    logger.info(
+                        "Rollback verification: manifest_total=%d, restored_total=%d",
+                        manifest_total,
+                        restored_total,
+                    )
+
+                    # Mark as not archived
+                    repo.set_runs_archived(session, [run.id], archived=False)
+                    session.commit()
+
+                result.success = True
+                click.echo(
+                    click.style(
+                        f"{'[DRY RUN] Would complete' if self.dry_run else 'Completed'} rollback for "
+                        f"workflow run {workflow_run_id}: restored={result.restored_counts}",
+                        fg="green",
+                    )
+                )
+
+            except Exception as e:
+                logger.exception("Failed to rollback workflow run %s", workflow_run_id)
+                result.error = str(e)
+                session.rollback()
+                click.echo(click.style(f"Rollback failed: {e}", fg="red"))
+
+        result.elapsed_time = time.time() - start_time
+        return result
+
+    def _get_workflow_run_repo(self) -> APIWorkflowRunRepository:
+        if self.workflow_run_repo is not None:
+            return self.workflow_run_repo
+
+        self.workflow_run_repo = DifyAPIRepositoryFactory.create_api_workflow_run_repository(
+            sessionmaker(bind=db.engine, expire_on_commit=False)
+        )
+        return self.workflow_run_repo
+
+    def _restore_table_records(
+        self,
+        session: Session,
+        table_name: str,
+        records: list[dict[str, Any]],
+    ) -> int:
+        """
+        Restore records to a table.
+
+        Uses INSERT ... ON CONFLICT DO NOTHING for idempotency.
+
+        Args:
+            session: Database session
+            table_name: Name of the table
+            records: List of record dictionaries
+
+        Returns:
+            Number of records actually inserted
+        """
+        if not records:
+            return 0
+
+        model = TABLE_MODELS.get(table_name)
+        if not model:
+            logger.warning("Unknown table: %s", table_name)
+            return 0
+
+        # Convert datetime strings back to datetime objects
+        converted_records = []
+        for record in records:
+            converted = self._convert_datetime_fields(record, model)
+            converted_records.append(converted)
+
+        # Use INSERT ... ON CONFLICT DO NOTHING for idempotency
+        stmt = pg_insert(model).values(converted_records)
```

Comment:
should we consider to do a schema mapping layer in case the schema has changed after the restore?